text/entities/0264-resource-and-entities.md

```diff
@@ -0,0 +1,852 @@
+rather than hard-coded rules that have to understand the nuance of when `host.id` influences `host.name`, e.g.
+
+### Users are expected to provide / prioritize "detectors" and determine which entity is "producing" or most-important for a signal
+
+The Resource Manager allows users to configure priority of Entity Detectors.
+
+### For an SDK - ALL telemetry should be associated with the same set of entities (resource labels)
+
+Resource Manager is responsible for resolving entities into a cohesive Resource that meets the same demands as Resource today.
+
+## Open Questions
+
+The following remain open questions:
+
+### How to attach Entity "bundle" information in Resource?
+
+The protocol today requires a raw grab bag of Attributes on Resource. We cannot break this going forward.  However, Entities represent a new mechanism of "bundling" attributes on Resource and interacting with these bundles.  We do not want this to bloat the protocol, nor do we want it to cause oddities.
+
+Going forward, we have set of options:
+
+- Duplicate attributes in `Entity` section of Resource.
+- Reference attributes of Resource in entity.
+- Only identify Entity id and keep attribute<->entity association out of band.
+- Extend Attribute on Resource so that we can track the entity type per Key-Value (across any attribute in OTLP).
+
+The third option prevents generic code from interacting with Resource and Entity without understanding the model of each.  The first keeps all usage of entity simple at the expense of duplicating information and the middle is awkward to interact with from an OTLP usage perspective. The fourth is violates our stability policy for OTLP.
+
+### How to deal with Resource/Entities whose lifecycle does not match the SDK?
+
+This proposal motivates a Resource Provider in the SDK whose job could include managing changes in entity lifetimes, but does not account for how these changes would be broadcast across TracerProvider, LogProvider, MeterProvider, etc.  That would be addressed in a follow on OTEP.
+
+### How to deal with Prometheus Compatibility for non-SDK telemetry?
+
+Today, [Prometheus compatibility](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md) relies on two key attributes in Resource: `service.name` and `service.instance.id`. These are not guaranteed to exist outside of OpenTelemetry SDK generation. While this question is not fully answered, we believe outlining identity in all resources within OpenTelemetry allows us to define a solution in the future while preserving compatibility with what works today.
+
+Here's a list of requirements for the solution:
+
+- Existing prometheus/OpenTelemetry users should be able to migrate from where they are today.
+- Prometheus should have a set of identifying attributes for their up and coming `info()` function and info metric features.
```

Comment:
I'd like to propose something different here. 

Let's first remember the motivation behind `target_info`. Resource attributes, before this OTEP, can be a gigantic array of information and not every attribute makes sense to become a metric label in the Prometheus world. Instead of adding all attributes to all metrics, which increases network bandwidth unnecessarily (almost all resources wouldn't be used at all), everything is encoded to `target_info` instead[[1](https://opentelemetry.io/docs/specs/otel/compatibility/prometheus_and_openmetrics/#:~:text=A%20Prometheus%20Info%20metric%20MUST%20be%20converted%20to%20an%20OTLP%20Non%2DMonotonic%20Sum%20unless%20it%20is%20the%20target%20info%20metric%2C%20which%20is%20used%20to%20populate%20resource%20attributes)]. 

Once `target_info` became a thing, PromQL Joins were the way to decode resource attributes from `target_info` and use them in queries with other metrics. Several learnings from this pattern eventually led to the creation of the `info()` function! The `info()` function is a massive UX improvement when compared to the usage of joins, but let's not forget the original intention of all of this: we want an efficient way to include just useful OTel attributes as metric labels.

I'm not sure if I'm missing something, but optimizing this proposal to make sure info() and identifying attributes work well together seems a bit like the [XY problem](https://www.google.com/url?sa=i&url=https%3A%2F%2Ftwitter.com%2Fsketchplanator%2Fstatus%2F1743988378863800608&psig=AOvVaw2hueW4eWExr97F9pYhF9_D&ust=1726763312621000&source=images&cd=vfe&opi=89978449&ved=0CBQQjRxqFwoTCOjy4-f0zIgDFQAAAAAdAAAAABAE).

--- 

This OTEP proposes a separation of resource attributes: identifying and descriptive attributes. Would it make sense to just add all identifying attributes as metric labels? 

Identifying attributes are immutable and, hopefully, a much shorter set when compared to the whole resource attribute set. This solves the problem of Prometheus receiving a gigantic set of labels that would never be used.

If descriptive attributes are still desired by users for query filters, would it make that the `prometheusexporter` and `prometheusremotewriteexporter` (otelcontribcol components) have configuration options that allow promoting particular descriptive attributes as labels? Additionally, we can propose that Prometheus /otlp endpoint can also promote particular descriptive attributes as labels.

Reply:
> If descriptive attributes are still desired by users for query filters, [...]
I think this statement is the core misconception, here. The identifying attributes are the shortest path to identify the entity technically, but it is not necessarily enough for a human to understand what that entity really is. It might be some extra technical ID on top of identifying service attributes that might not even make sense to the end user receiving the data. I can share the example that brought me to this discussion if it helps.

Reply:
Agree with @feldentm-SAP. I think that bringing OTel resource attribute to Prometheus label promotion into this proposal is a distraction. As I see it, whether/how to promote resource attributes should be layered on top of basic OTel/Prometheus interop, which is what should be covered in the Entity proposal.

Reply:
Hmmm, okay. My understanding was that it would be rare for people to want descriptive attributes in their queries.

Would it make sense to list what you predict to be common identifying and descriptive attributes in the proposal?

Reply:
> Would it make sense to list what you predict to be common identifying and descriptive attributes in the proposal?

Won't this be entirely domain dependent?

Reply:
> Won't this be entirely domain dependent?

Yes, I think you're correct. Maybe it's too much to add to the proposal. I'm just hoping for a prediction so we can compare it to our [current recommendations](https://prometheus.io/docs/guides/opentelemetry/#promoting-resource-attributes).

Reply:
Again, the concepts of identifying OTel resource attributes and which to promote to Prometheus metric labels should be kept orthogonal to each other.

Reply:
Just for clarification: I expect that identifying attributes will overlap, i.e. belong to multiple entities. Because we have requirements to express sub resources. The implications of this vary depending on Prometheus' role in the observability stack. If it is used as a routing component, it should pretty much behave like an OTel collector and all available attributes must be passed through it. If it is used as data sink, a consumer might want to filter some irrelevant labels before. However, I'm not aware of this happening a lot and would, hence, not let it guide design decisions, right?

Reply:
I think this is an interesting proposal, but I think it's important to understand why resource exists in otel in the first place.

I wrote a bit about how it's used [here](https://docs.google.com/document/d/1Xd1JP7eNhRpdz1RIBLeA1_4UYPRJaouloAYqldCeNSc/edit).

A shortened version:

- It must identify an entity that is producing telemetry.
- It should allow users to determine where that entity resides within their infrastructure.

Prometheus metric labels *can* do this as well. But a few things I don't have in that paper that's important for OpenTelemetry:

- Resource provides "guaranteed" mechanism for correlating o11y signals by "source".
- OpenTelemetry pipelines (OTLP) leverage resource as a default bundling / aggregation boundary for signals.

When we speak of how things encode in prometheus, there's a few questions I have:

1. How will users of prometheus interact with other signals?  E.g. how can I go from a an alert on a timeseries to finding relevant logs/traces from the time-period of the anamolous behavior?  Does preserving the semantic of "resource" make this easier?
2. When federating Prometheus or interacting with OTLP, should we be able to "round trip" this semantic information both in & out?  I.e. should we be able to export OTLP from prometheus that looks (as close as possible) to the OTLP that created the timeseries in Prometheus to begin with?

In my mind - I would love if users of Prometheus and other backends for OTLP signals can consistently understand "resource.{label}" means the same thing between the signals, and can design their system around this.

In terms of our discussion, let's outline things I'm implicitly hearing agreement to, or that I think are inputs into the design space:

- `Info()` seems needed to deal with descriptive attributes, no matter what.  The proposals I've seen for `Info`, I think, solve this issue well.
- Having an arbitrary number of labels to `join` Info metrics is problematic from performance and implementation persective (@aknuds1 let me know if I'm reading too much into this).
- OpenTelemetry (today) allows an aribitrary number of identifying attributes.  Unfortunately, users are placing descriptive attributes in that location today, because of the usefulness of "resource" as a concept for x-telemetry correlation and labeling.
- OpenTelemetry *only* guarantees `job/instance` labels for SDK where we can ensure a `service.name` and `service.instance.id` are required (although that's not fully consistent across our SDKs yet, as it's "new"-ish requirement).
- Entities proposal gives us the ability to split identifying + descriptive attributes in Resource.  Additionally, we will have some notion of types of resource.  

Given this requirements, let's try to answer the two questions I listed above.  Looking forward to the discussion later today.

Reply:
@jsuereth 

> How will users of prometheus interact with other signals? E.g. how can I go from a an alert on a timeseries to finding relevant logs/traces from the time-period of the anamolous behavior? Does preserving the semantic of "resource" make this easier?

I'm not the biggest expert on this, but in my personal experience exemplars are useful for finding traces corresponding to metrics. In the OTel sphere, I think that knowing the identifying resource attributes (ideally via Prometheus metadata?) would be really helpful in finding other signals from the same producer. Agree?

> When federating Prometheus or interacting with OTLP, should we be able to "round trip" this semantic information both in & out? I.e. should we be able to export OTLP from prometheus that looks (as close as possible) to the OTLP that created the timeseries in Prometheus to begin with?

I haven't given much thought to round tripping, but ideally it should be possible to re-create the OTLP losslessly. I figure that storing OTel resource attributes as metadata ought to facilitate this scenario. Make sense?

> Info() seems needed to deal with descriptive attributes, no matter what. The proposals I've seen for Info, I think, solve this issue well.

Yeah, the `info` [MVP](https://github.com/prometheus/prometheus/pull/14495) which will hopefully be approved, basically acts like a smart join query against `target_info`. So it depends on `instance` and `job` being identifying labels, plus all descriptive attributes being labels of `target_info`.

> Having an arbitrary number of labels to join Info metrics is problematic from performance and implementation persective (@aknuds1 let me know if I'm reading too much into this).

The current version of `info`, which doesn't require any new Prometheus metadata capabilities, depends on `instance` and `job` being _the_ identifying `target_info` labels. If Prometheus were to have persisted metadata, `info` could handle arbitrary identifying labels, but too many would still cause metrics to have too many labels (could hit limits) etc. It's better to keep it simple :)

Reply:
> I'm not the biggest expert on this, but in my personal experience exemplars are useful for finding traces corresponding to metrics.

OpenTelemetry provides (at least) two types of correlation:

- "Causal" correlation.  You *know* that "A" leads to "B".  We do this through (trace) id propagation and Exemplars are a means to track participants in the "cause" of a metric being the way it is.
- "Source" correlation.  This is what "Resource" does.  What system/component created "signal A"?  Is that the same component that's also creating problematic signal "B"?

They're different goals and both are viable/useful.  We can't only rely on a (trace) IDs for causal relationships, just like we can't only rely on causal relationships for o11y use cases.

Reply:
> In the OTel sphere, I think that knowing the identifying resource attributes (ideally via Prometheus metadata?) would be really helpful in finding other signals from the same producer. Agree?

Yes.

> I haven't given much thought to round tripping, but ideally it should be possible to re-create the OTLP losslessly. I figure that storing OTel resource attributes as metadata ought to facilitate this scenario. Make sense?

Yes, I think metadata should be in the design considerations here.

My main concern is the details on how this would work, and I'm relying on @aknuds1's opinion here for viability.

Reply:
Based on the in person proposal, I've updated this section. TL:DR;

- We continue to leverage `target_info` and `Resource.attributes` as we do today.  
- We will work together to design entities and Prometheus features to better the experience of O11y of OpenTelemetry + Prometheus.  We can progress with this proposal and hash out the prometheus<->otel design as we go.